File: backend/internal/crypto/chacha.go
package crypto

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"strings"

	"golang.org/x/crypto/chacha20poly1305"
)

// Encrypt encrypts plaintext using ChaCha20-Poly1305 with a 32-byte key.
// Returns "base64(nonce):base64(ciphertext+auth_tag)"
func Encrypt(key, plaintext []byte) (string, error) {
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return "", fmt.Errorf("create cipher: %w", err)
	}

	nonce := make([]byte, aead.NonceSize())
	if _, err := rand.Read(nonce); err != nil {
		return "", fmt.Errorf("generate nonce: %w", err)
	}

	ciphertext := aead.Seal(nil, nonce, plaintext, nil)

	return base64.RawStdEncoding.EncodeToString(nonce) + ":" +
		base64.RawStdEncoding.EncodeToString(ciphertext), nil
}

// Decrypt decrypts a value produced by Encrypt.
// Expects "base64(nonce):base64(ciphertext+auth_tag)"
func Decrypt(key []byte, encoded string) ([]byte, error) {
	parts := strings.SplitN(encoded, ":", 2)
	if len(parts) != 2 {
		return nil, fmt.Errorf("invalid encrypted format")
	}

	nonce, err := base64.RawStdEncoding.DecodeString(parts[0])
	if err != nil {
		return nil, fmt.Errorf("decode nonce: %w", err)
	}

	ciphertext, err := base64.RawStdEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("decode ciphertext: %w", err)
	}

	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, fmt.Errorf("create cipher: %w", err)
	}

	plaintext, err := aead.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}

	return plaintext, nil
}

// GenerateSymmetricKey generates a random 32-byte ChaCha20 key.
func GenerateSymmetricKey() ([]byte, error) {
	key := make([]byte, chacha20poly1305.KeySize)
	if _, err := rand.Read(key); err != nil {
		return nil, fmt.Errorf("generate key: %w", err)
	}
	return key, nil
}

File: cli/src/crypto/chacha.rs
use anyhow::{anyhow, Context, Result};
use base64::{engine::general_purpose::STANDARD as B64, Engine};
use ring::aead::{Aad, LessSafeKey, Nonce, UnboundKey, CHACHA20_POLY1305};
use ring::rand::{SecureRandom, SystemRandom};

/// Encrypt plaintext with ChaCha20-Poly1305.
/// Returns "base64(nonce):base64(ciphertext+auth_tag)"
pub fn encrypt(key: &[u8; 32], plaintext: &[u8]) -> Result<String> {
    let rng = SystemRandom::new();
    let mut nonce_bytes = [0u8; 12];
    rng.fill(&mut nonce_bytes).map_err(|_| anyhow!("generate nonce failed"))?;

    let unbound =
        UnboundKey::new(&CHACHA20_POLY1305, key).map_err(|_| anyhow!("invalid key"))?;
    let sealing_key = LessSafeKey::new(unbound);

    let mut in_out = plaintext.to_vec();
    sealing_key
        .seal_in_place_append_tag(
            Nonce::assume_unique_for_key(nonce_bytes),
            Aad::empty(),
            &mut in_out,
        )
        .map_err(|_| anyhow!("encryption failed"))?;

    Ok(format!("{}:{}", B64.encode(nonce_bytes), B64.encode(&in_out)))
}

/// Decrypt a value produced by encrypt().
/// Expects "base64(nonce):base64(ciphertext+auth_tag)"
pub fn decrypt(key: &[u8; 32], encoded: &str) -> Result<Vec<u8>> {
    let parts: Vec<&str> = encoded.splitn(2, ':').collect();
    if parts.len() != 2 {
        return Err(anyhow!("invalid encrypted format"));
    }

    let nonce_bytes: [u8; 12] = B64
        .decode(parts[0])
        .context("decode nonce")?
        .try_into()
        .map_err(|_| anyhow!("nonce wrong length"))?;

    let mut ciphertext = B64.decode(parts[1]).context("decode ciphertext")?;

    let unbound =
        UnboundKey::new(&CHACHA20_POLY1305, key).map_err(|_| anyhow!("invalid key"))?;
    let opening_key = LessSafeKey::new(unbound);

    let plaintext = opening_key
        .open_in_place(
            Nonce::assume_unique_for_key(nonce_bytes),
            Aad::empty(),
            &mut ciphertext,
        )
        .map_err(|_| anyhow!("decryption failed — wrong key or corrupted data"))?;

    Ok(plaintext.to_vec())
}
